io/io/src/TStreamerInfoActions.cxx

```diff
@@ -441,6 +568,259 @@ namespace TStreamerInfoActions
    }
 
 
+   /** Direct copy of code from TStreamerInfo::WriteBufferAux,
```

Comment:
As a side note, the idea of the Action is to factor these long blocks to remove any if statements whose answer is known at construction time.  For example those at line 583 or 605 should (eventually) lead to 2 more version of this functions (possibly build by composition to avoid code duplication) where each version contains only the code path needed and no (well less) conditional statements.

Reply:
Yes, I agree.

As you can see, I keep possibility to use same actions for binary- and text-based buffers.
Should one try to use these actions also for binary I/O?

Also very principal question - should we try to avoid virtual methods inside actions?
This can be achieved, if we can really split binary-based and text-based TBuffer classes and use respective methods directly. But this could have many side-effects.

Reply:
> Should one try to use these actions also for binary I/O?

Yes, especially in some work goes into optimization (i.e. splitting the function).

> should we try to avoid virtual methods inside actions?

Absolutely, that is the main advantage of splitting (even if just by templating) text and binary functions.

Thanks.

Reply:
>>    should we try to avoid virtual methods inside actions?

> Absolutely, that is the main advantage of splitting (even if just by templating) text and binary functions.

My current idea - introduce TBufferText class which is derived from TBuffer and defines all necessary methods to work with text-based streamers. And all JSON/XML/SQL buffers derive from it. Problem - XML and SQL buffers are inherited from TBufferFile, which could be difficult to brake. 

But if it would work, TBufferFile interfaces can be used exclusively for binary operations and TBufferText exclusively for text-based. Of course, all virtual methods for TBuffer will be implemented for both classes. I will check if it can be done this way.

Reply:
> But if it would work, TBufferFile interfaces can be used exclusively for binary operations and TBufferText exclusively for text-based

if the inheritance from TBufferFile is still needed for TBufferXML, I think we can still pull it of, as
```
TBufferFile::WriteInt(...);
TBufferText::WriteInt(...);
```
would both call the right thing without virtual call ... humm but they actually require casting the input (TBuffer) into the expected type (TBufferFile or TBufferText), we have to make sure the cast is not 'eating' all the gain :)

Reply:
> we have to make sure the cast is not 'eating' all the gain 

I hope, static_cast should cost nothing, but it should be checked precisely.